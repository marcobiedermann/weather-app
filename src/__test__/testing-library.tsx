import { RenderOptions, render } from '@testing-library/react';
import { ReactElement, ReactNode } from 'react';
import { I18nextProvider } from 'react-i18next';
import { MemoryRouter } from 'react-router-dom';
import i18n from './i18n';

interface ProvidersProps {
  children: ReactNode;
}

function Providers(props: ProvidersProps) {
  const { children } = props;

  return (
    <I18nextProvider i18n={i18n}>
      <MemoryRouter>{children}</MemoryRouter>
    </I18nextProvider>
  );
}

function customRender(ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) {
  return render(ui, {
    wrapper: Providers,
    ...options,
  });
}

export { customRender as render };
