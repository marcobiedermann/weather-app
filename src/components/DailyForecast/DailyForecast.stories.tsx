import type { Meta, StoryObj } from '@storybook/react';
import DailyForecast from '.';

const meta: Meta<typeof DailyForecast> = {
  args: {
    list: [
      {
        dt: 1687086000,
        temp: {
          max: 26.22,
          min: 13.87,
        },
        weather: [
          {
            description: 'sky is clear',
            icon: '01d',
            id: 800,
          },
        ],
      },
    ],
  },
  component: DailyForecast,
};

type Story = StoryObj<typeof DailyForecast>;

const Default: Story = {};

export { Default };
export default meta;
