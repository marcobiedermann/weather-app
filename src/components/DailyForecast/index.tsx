import { format, formatISO, fromUnixTime } from 'date-fns';
import { useTranslation } from 'react-i18next';
import { DAY_OF_WEEK } from '../../utils/date';

interface Temp {
  max: number;
  min: number;
}

interface Weather {
  description: string;
  icon: string;
  id: number;
}
interface List {
  dt: number;
  temp: Temp;
  weather: Weather[];
}
interface DailyForecastProps {
  list: List[];
}

function DailyForecast(props: DailyForecastProps): JSX.Element {
  const { list } = props;
  const { t } = useTranslation();

  return (
    <table>
      <tbody>
        {list.map((day) => {
          const dateTime = fromUnixTime(day.dt);

          return (
            <tr key={day.dt}>
              <td>
                <time
                  dateTime={formatISO(dateTime, {
                    representation: 'date',
                  })}
                >
                  {format(dateTime, DAY_OF_WEEK)}
                </time>
              </td>
              <td>
                {day.weather.map((entry) => (
                  <img
                    src={`https://openweathermap.org/img/wn/${entry.icon}.png`}
                    alt={entry.description}
                    key={entry.id}
                  />
                ))}
              </td>
              <td>{t('intlNumberCelsius', { val: day.temp.min })}</td>
              <td>{t('intlNumberCelsius', { val: day.temp.max })}</td>
            </tr>
          );
        })}
      </tbody>
    </table>
  );
}

export type { DailyForecastProps };
export default DailyForecast;
