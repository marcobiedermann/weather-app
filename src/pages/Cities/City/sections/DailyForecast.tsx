import DailyForecast from '../../../../components/DailyForecast';
import Error from '../../../../components/Error';
import Loader from '../../../../components/Loader';
import { useDailyForecast } from '../../../../hooks';

interface DailyForecastSectionProps {
  cityId: number;
  settings: any;
}

function DailyForecastSection(props: DailyForecastSectionProps) {
  const { cityId, settings } = props;
  const { data: dailyForecast, error, isError, isLoading } = useDailyForecast(cityId, settings);

  if (isError) {
    return <Error message={error.message} />;
  }

  if (isLoading) {
    return <Loader />;
  }

  return (
    <section>
      <h3>10-Day Forecast</h3>
      <DailyForecast list={dailyForecast.list.slice(0, 10)} />
    </section>
  );
}

export type { DailyForecastSectionProps };
export default DailyForecastSection;
