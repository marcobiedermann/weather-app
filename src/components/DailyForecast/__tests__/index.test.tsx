import { describe, expect, it } from 'vitest';
import DailyForecast from '..';
import { render } from '../../../__test__/testing-library';

describe('components/DailyForecast', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<DailyForecast list={[]} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
