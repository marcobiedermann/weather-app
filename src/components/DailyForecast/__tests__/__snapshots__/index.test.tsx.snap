// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/DailyForecast > renders correctly 1`] = `
<DocumentFragment>
  <table>
    <tbody />
  </table>
</DocumentFragment>
`;
